src/nlsolve/newton.jl

```diff
@@ -64,27 +49,35 @@ Equations II, Springer Series in Computational Mathematics. ISBN
 """
 @muladd function compute_step!(nlsolver::NLSolver{<:NLNewton,false}, integrator)
   @unpack uprev,t,p,dt,opts = integrator
-  @unpack z,tmp,γ,α,cache = nlsolver
+  @unpack alg,z,tmp,γ,α,cache = nlsolver
   @unpack tstep,W,invγdt = cache
 
   f = nlsolve_f(integrator)
   isdae = f isa DAEFunction
 
-  if !isdae
-    mass_matrix = integrator.f.mass_matrix
-  end
-
   if isdae
     ustep = @. uprev + z
     dustep = @. (tmp + α * z) * invγdt
     ztmp = f(dustep, ustep, p, t)
   else
-    ustep = @. tmp + γ * z
-    if mass_matrix === I
-      ztmp = (dt .* f(ustep, p, tstep) .- z) .* invγdt
+    mass_matrix = integrator.f.mass_matrix
+    if isodemultistep(unwrap_alg(integrator, true))
```

Comment:
this will cause problems in SDEs.

Reply:
How can I make it StochasticDiffEq friendly?

Reply:
I think StochasticDiffEq has `unwrap_alg` overload, too.

Reply:
Oh nevermind, https://github.com/SciML/OrdinaryDiffEq.jl/pull/1160/files#diff-aac6396f64983b0f4a67297163d3155aR572 that will catch it. But I still think there should be two caches so each algorithm picks which one to use. That would make general linear methods possible.